File: app/src/main/java/com/softwareverde/tidyduck/most/FunctionBlock.java
package com.softwareverde.tidyduck.most;

import com.softwareverde.mostadapter.Modification;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class FunctionBlock {
    private Long _id;
    private String _mostId;
    private String _kind = "Proprietary";
    private String _name;
    private String _description;
    private String _release;
    private Date _lastModifiedDate;
    private Author _author;
    private Company _company;
    private String _access;
    private boolean _isSink;
    private boolean _isSource;
    private boolean _isDeleted;
    private Date _deletedDate;
    private boolean _isPermanentlyDeleted;
    private Date _permanentlyDeletedDate;
    private boolean _isApproved;
    private Long _approvalReviewId;
    private boolean _isReleased;
    private Long _baseVersionId;
    private Long _priorVersionId;
    private Long _creatorAccountId;
    private List<Modification> _modifications = new ArrayList<>();
    private List<MostInterface> _mostInterfaces = new ArrayList<>();

    public Long getId() {
        return _id;
    }

    public void setId(Long id) {
        _id = id;
    }

    public String getMostId() {
        return _mostId;
    }

    public void setMostId(String mostId) {
        _mostId = mostId;
    }

    public String getKind() {
        return _kind;
    }

    public void setKind(String kind) {
        _kind = kind;
    }

    public String getName() {
        return _name;
    }

    public void setName(String name) {
        _name = name;
    }

    public String getDescription() {
        return _description;
    }

    public void setDescription(String description) {
        _description = description;
    }

    public String getRelease() {
        return _release;
    }

    public void setRelease(String release) {
        _release = release;
    }

    public Date getLastModifiedDate() {
        return _lastModifiedDate;
    }

    public void setLastModifiedDate(Date lastModifiedDate) {
        _lastModifiedDate = lastModifiedDate;
    }

    public Author getAuthor() {
        return _author;
    }

    public void setAuthor(Author author) {
        _author = author;
    }

    public Company getCompany() {
        return _company;
    }

    public void setCompany(Company company) {
        _company = company;
    }

    public void setAccess(String access) {
        _access = access;
    }

    public String getAccess() {
        return _access;
    }

    public boolean isSink() {
        return _isSink;
    }

    public void setIsSink(final boolean sink) {
        _isSink = sink;
    }

    public boolean isSource() {
        return _isSource;
    }

    public void setIsSource(final boolean source) {
        _isSource = source;
    }

    public boolean isDeleted() {
        return _isDeleted;
    }

    public void setIsDeleted(final boolean deleted) {
        _isDeleted = deleted;
    }

    public Date getDeletedDate() {
        return _deletedDate;
    }

    public void setDeletedDate(final Date deletedDate) {
        _deletedDate = deletedDate;
    }

    public boolean isPermanentlyDeleted() {
        return _isPermanentlyDeleted;
    }

    public void setIsPermanentlyDeleted(final boolean permanentlyDeleted) {
        _isPermanentlyDeleted = permanentlyDeleted;
    }

    public Date getPermanentlyDeletedDate() {
        return _permanentlyDeletedDate;
    }

    public void setPermanentlyDeletedDate(final Date permanentlyDeletedDate) {
        _permanentlyDeletedDate = permanentlyDeletedDate;
    }

    public boolean isApproved() {
        return _isApproved;
    }

    public void setIsApproved(final boolean approved) {
        _isApproved = approved;
    }

    public Long getApprovalReviewId() { return _approvalReviewId; }

    public void setApprovalReviewId(final Long approvalReviewId) { _approvalReviewId = approvalReviewId; }

    public boolean isReleased() {
        return _isReleased;
    }

    public void setIsReleased(boolean released) {
        _isReleased = released;
    }

    public Long getBaseVersionId() {
        return _baseVersionId;
    }

    public void setBaseVersionId(Long baseVersionId) {
        _baseVersionId = baseVersionId;
    }

    public Long getPriorVersionId() {
        return _priorVersionId;
    }

    public void setPriorVersionId(Long priorVersionId) {
        _priorVersionId = priorVersionId;
    }

    public Long getCreatorAccountId() { return _creatorAccountId; }

    public void setCreatorAccountId(final Long creatorAccountId) { _creatorAccountId = creatorAccountId; }

    public List<Modification> getModifications() {
        return _modifications;
    }

    public void addModification(final Modification modification) {
        _modifications.add(modification);
    }

    public List<MostInterface> getMostInterfaces() {
        return new ArrayList<>(_mostInterfaces);
    }

    public void addMostInterface(MostInterface mostInterface) {
        _mostInterfaces.add(mostInterface);
    }

    public void setMostInterfaces(final List<MostInterface> mostInterfaces) {
        _mostInterfaces = new ArrayList<>(mostInterfaces);
    }
}

File: www/js/classes/function-block.js
class FunctionBlock {
    static fromJson(json) {
        const functionBlock = new FunctionBlock();

        const author = new Author();
        author.setId(json.authorId);
        author.setName(json.authorName);

        const company = new Company();
        company.setId(json.companyId);
        company.setName(json.companyName);

        functionBlock.setId(json.id);
        functionBlock.setMostId(json.mostId);
        functionBlock.setKind(json.kind);
        functionBlock.setName(json.name);
        functionBlock.setDescription(json.description);
        functionBlock.setLastModifiedDate(json.lastModifiedDate);
        functionBlock.setReleaseVersion(json.releaseVersion);
        functionBlock.setBaseVersionId(json.baseVersionId);
        functionBlock.setPriorVersionId(json.priorVersionId);
        functionBlock.setIsReleased(json.isReleased);
        functionBlock.setIsApproved(json.isApproved);
        functionBlock.setApprovalReviewId(json.approvalReviewId);
        functionBlock.setHasApprovedParent(json.hasApprovedParent);
        functionBlock.setAuthor(author);
        functionBlock.setCompany(company);
        functionBlock.setCreatorAccountId(json.creatorAccountId);
        functionBlock.setAccess(json.access);
        functionBlock.setIsSource(json.isSource);
        functionBlock.setIsSink(json.isSink);
        functionBlock.setIsDeleted(json.isDeleted);
        functionBlock.setDeletedDate(json.deletedDate);

        return functionBlock;
    }

    static toJson(functionBlock) {
        const jsonFunctionBlock = {
            id:                 functionBlock.getId(),
            mostId:             formatHex(functionBlock.getMostId()),
            kind:               functionBlock.getKind(),
            name:               functionBlock.getName(),
            description:        functionBlock.getDescription(),
            lastModifiedDate:   functionBlock.getLastModifiedDate(),
            releaseVersion:     functionBlock.getReleaseVersion(),
            baseVersionId:      functionBlock.getBaseVersionId(),
            priorVersionId:     functionBlock.getPriorVersionId(),
            creatorAccountId:   functionBlock.getCreatorAccountId(),
            isReleased:         functionBlock.isReleased(),
            isApproved:         functionBlock.isApproved(),
            access:             functionBlock.getAccess(),
            isSource:           functionBlock.isSource(),
            isSink:             functionBlock.isSink(),
        };
        const author = (functionBlock.getAuthor() || new Author());
        const company = (functionBlock.getCompany() || new Company());
        if (author.getId() > 0) {
            jsonFunctionBlock.authorId = author.getId();
        }
        if (company.getId() > 0) {
            jsonFunctionBlock.companyId = company.getId();
        }
        return jsonFunctionBlock;
    }

    constructor() {
        this._id                    = null;
        this._mostId                = null;
        this._kind                  = "";
        this._name                  = "";
        this._description           = "";
        this._lastModifiedDate      = "";
        this._releaseVersion        = "";
        this._author                = null;
        this._company               = null;
        this._access                = "";
        this._isSource              = false;
        this._isSink                = false;
        this._versionsJson          = null;
        this._isReleased            = null;
        this._isApproved            = null;
        this._approvalReviewId      = null;
        this._hasApprovedParent     = false;
        this._priorVersionId        = null;
        this._baseVersionId         = null;
        this._creatorAccountId      = null;
        this._isDeleted             = false;
        this._deletedDate           = null;

        this._interfaces            = [];
    };

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setMostId(mostId) {
        this._mostId = mostId;
    }

    getMostId() {
        return this._mostId;
    }

    setKind(kind) {
        this._kind = kind;
    }

    getKind() {
        return this._kind;
    }

    setName(name) {
        this._name = name;
    }

    getName() {
        return this._name;
    }

    setDescription(description) {
        this._description = description;
    }

    getDescription() {
        return this._description;
    }

    setLastModifiedDate(lastModifiedDate) {
        this._lastModifiedDate = lastModifiedDate;
    }

    getLastModifiedDate() {
        return this._lastModifiedDate;
    }

    setReleaseVersion(releaseVersion) {
        this._releaseVersion = releaseVersion;
    }

    getReleaseVersion() {
        return this._releaseVersion;
    }

    setAuthor(author) {
        this._author = author;
    }

    getAuthor() {
        return this._author;
    }

    setCompany(company) {
        this._company = company;
    }

    getCompany() {
        return this._company;
    }

    getInterfaces() {
        return this._interfaces;
    }

    setIsSource(isSource) {
        this._isSource = isSource;
    }

    isSource() {
        return this._isSource;
    }

    setIsSink(isSink) {
        this._isSink = isSink;
    }

    isSink() {
        return this._isSink;
    }

    setAccess(access) {
        this._access = access;
    }

    getAccess() {
        return this._access;
    }

    setVersionsJson(versionsJson) {
        this._versionsJson = versionsJson;
    }

    getVersionsJson() {
        return this._versionsJson;
    }

    setBaseVersionId(baseVersionId) {
        this._baseVersionId = baseVersionId;
    }

    getBaseVersionId() {
        return this._baseVersionId;
    }

    setPriorVersionId(priorVersionId) {
        this._priorVersionId = priorVersionId;
    }

    getPriorVersionId() {
        return this._priorVersionId;
    }

    setIsReleased(isReleased) {
        this._isReleased = isReleased;
    }

    isReleased() {
        return this._isReleased;
    }

    setIsApproved(isApproved) {
        this._isApproved = isApproved;
    }

    isApproved() {
        return this._isApproved;
    }

    setApprovalReviewId(approvalReviewId) {
        this._approvalReviewId = approvalReviewId;
    }

    getApprovalReviewId() {
        return this._approvalReviewId;
    }

    setHasApprovedParent(hasApprovedParent) {
        this._hasApprovedParent = hasApprovedParent;
    }

    hasApprovedParent() {
        return this._hasApprovedParent;
    }

    setCreatorAccountId(creatorAccountId) {
        this._creatorAccountId = creatorAccountId;
    }

    getCreatorAccountId() {
        return this._creatorAccountId;
    }

    setIsDeleted(isDeleted) {
        this._isDeleted = isDeleted;
    }

    isDeleted() {
        return this._isDeleted;
    }

    setDeletedDate(deletedDate) {
        this._deletedDate = deletedDate;
    }

    getDeletedDate() {
        return this._deletedDate;
    }

    getDisplayVersion() {
        if (this._isReleased) {
            return this._releaseVersion;
        }
        return this._releaseVersion + "-" + this._id;
    }
}
